Refactor functions. Add realtime data sources.

adamlwgriffiths opened the pull request:
Use specific versions of C3/D3.js.
Use CDN versions of C3/D3.js.
Remove unused ID setting.
Change "columns" to be "data". This allows us to send
rows, columns, or json in a dictionary.
Change "options" to be "text" instead of "calculated".
Add a "flow" boolean setting. Setting to true will use
chart.flow instead of chart.load.
Rename variable "myElement" to "element".
Change element class from "testing" to "c3-chart-container".
Move padding calculations into a function.
Add "objectStringToJson" function which converts "obj.toString()"
to proper JSON.
Parse "options" as either "obj.toString()" or JSON.
Send empty columns during rendering. Calculated fields are not
calculated until "onCalculatedValueChanged", which means that
putting datasource references in here will cause an error.
Handle "extend(options)" safely, without this an invalid option
will cause empty C3 widgets to be built.
Move chart creation into it's own function outside of render.
Check for chart type changes specifically and call transform.
Check for options changes and recreate the chart if options change.
Handle both load and flow methods of pushing data.
Remove "type" change check in "onCalculatedValueChanged". Type is
not a calculated value and won't be sent here.
Remove evals.
Remove debug prints.
Update README example.

adamlwgriffiths committed:
Refactor functions. Add realtime data sources.

Use specific versions of C3/D3.js.
Use CDN versions of C3/D3.js.
Remove unused ID setting.
Change "columns" to be "data". This allows us to send
rows, columns, or json in a dictionary.
Change "options" to be "text" instead of "calculated".
Add a "flow" boolean setting. Setting to true will use
chart.flow instead of chart.load.
Rename variable "myElement" to "element".
Change element class from "testing" to "c3-chart-container".
Move padding calculations into a function.
Add "objectStringToJson" function which converts "obj.toString()"
to proper JSON.
Parse "options" as either "obj.toString()" or JSON.
Send empty columns during rendering. Calculated fields are not
calculated until "onCalculatedValueChanged", which means that
putting datasource references in here will cause an error.
Handle "extend(options)" safely, without this an invalid option
will cause empty C3 widgets to be built.
Move chart creation into it's own function outside of render.
Check for chart type changes specifically and call transform.
Check for options changes and recreate the chart if options change.
Handle both load and flow methods of pushing data.
Remove "type" change check in "onCalculatedValueChanged". Type is
not a calculated value and won't be sent here.
Remove evals.
Remove debug prints.
Update README example.

adamlwgriffiths commented:
Pretty big changes.
Supporting `columns`, `rows`, and `json` requires `data` field to be a dictionary/object instead of the previous hard coded `columns` value.
This could be done by providing a setting for each, but it's not as flexible.
The `data` setting allows you to provide more than just the column data to the `load` and `flow` functions - such as length, to, etc (see here: http://c3js.org/reference.html#api-flow).

The change to `options` from `calculated` to `text` may be annoying as text has a small text box and no large edit mode (JS Editor).
The issue with `options` being `calculated`, is that on each save the options would trigger an `onCalculatedValueChanged` with the settingName `options` even if nothing in it was changed.
You cannot safely compare this to the `currentSettings.options` if it contains datasource references or is a function, as the `currentSettings` version will contain the verbatim text, and `newValue` will contain the calculated result.
Which in turn means that every time you accept the settings, the graph will be completely redrawn even if nothing in `options` changed. Although this isn't that big of a deal.
There may also be a use-case for getting the options from a datasource.
If you want me to revert options to `calculated`, let me know.

Lobaro commented:
Thanks for sharing @ both of you.

I think it would be better to set the options also as calculated. Reason is that you can add custom legend driven by datasource (as you already mentioned). 2nd i don't get to point of your json regex function. wasn't able to set more than 1 option with it and removed it.

adamlwgriffiths commented:
The regex is to provide support for both serialized objects, and json format (Trust Javascript to invent 2 notations for the same thing).

```
obj.toString()
{a: "b"}

JSON.stringify(obj)
{"a": "b"}
```

If you attempt to `JSON.parse` a `toString`'ed object, it will fail. JSON insists on having " around the keys.
See this question here: https://stackoverflow.com/questions/9036429/convert-object-string-to-json
So the regex converts serialized objects to JSON format by adding those quotes to any dictionary/hash/map key that doesn't have them.
The calculated fields dont specify if the data is JSON (which may be received from a datasource), or object notation, which may be entered manually. I found myself entering both and felt that if I was doing so, it was important to support both.
I also really wanted to get rid of those evals, because they could trigger any number of insecure actions.

I'm not sure what you mean by "wasn't able to set more than 1 option with it and removed it."
I tested the data setting with JSON, serialized objects, and mixes of the two, and didn't have any issues.

I'll convert the options to calculated tomorrow.

DJWassink commented:
Thanks again for another PR, currently I am pretty busy so I will try to take a better look at it asap.

Nice stuff btw about the stringify/toString didn't know about that!

Lobaro commented:
Thanks for explanation, makes sense now to me as embedded guy :-) 
See it in action here: http://www.lobaro.com/coap-over-6lowpan-introduction
Will try to contribute a responsive design update soon.

adamlwgriffiths commented:
I've converted the options field back to calculated.

To ensure that graphs don't get recreated each time you accept the options, I have to store a calculated copy of the options to compare to (verbatim options may contain unresolved 'datasource' references).
I changed the `.toString` to `JSON.stringify` as `.toString` only seems to serialise the top level. It's more like a description than an actual serialization. Again, Javascript. Go figure.

The second issue is that empty calculated fields do not trigger `onCalculatedValueChanged`.
This means that there needs to be a second check inside `onSettingsChanged` to check if the new options differ from the current options AND the new options are empty.

All of this makes the code more fragile / cumbersome. I'm not sure if it's worth doing. I'm happy to revert it if desired.

I also have a "workaround" for an issue in C3 (https://github.com/masayuki0812/c3/issues/1097) where data pushed using `chart.flow` accumulates while the tab is not visible.
This bug has 2 parts.
1. When the tab becomes visible again, the old data will not be trimmed correctly (the `data.length` value is not handled properly).
2. The data accumulates while the tab is not visible and can lead to out of memory issues.

The "workaround" I have involves buffering data while the tab is not visible, but there are 2 issues.
1. It's again, a fiddly, cumbersome, work around.
2. I've still seen the tabs go unresponsive in chrome.
3. The graph becomes inconsistent and displays old data next to new (the buffer is limited size and once the maximum is hit, old data is dropped).

I tried destroying the graph and re-creating it in this situation, but have been unsuccessful.

I have no included this workaround in this PR as I feel it is a seperate issue, I just want to mention it.
The source tree for this "workaround" can be found here: https://github.com/someones/freeboard-c3-charts/tree/buffering

adamlwgriffiths commented:
Actually, give me a second. The new options code isn't right.

adamlwgriffiths commented:
Ok, it "should be fixed".
As you can see, the change to options being calculated is quite cumbersome and error prone.
I'm not sure it's worth the trouble.

adamlwgriffiths commented:
Ok. Because of the fragility of that code, I've 'force' backed it out and put it in a new branch. I've made a new PR for that change #3.
